finality-aleph/src/data_io/data_store.rs

```diff
@@ -0,0 +1,655 @@
+        let chain_info_provider = CachedChainInfoProvider::new(client.clone(), Default::default());
+
+        let highest_finalized_num = status.finalized_number;
+        (
+            DataStore {
+                next_free_id: 0,
+                pending_proposals: HashMap::new(),
+                event_triggers: HashMap::new(),
+                pending_messages: BTreeMap::new(),
+                chain_info_provider,
+                requested_blocks_cache: LruCache::new(config.requested_block_cache_capacity),
+                available_proposals_cache: LruCache::new(config.available_proposals_cache_capacity),
+                num_triggers_registered_since_last_pruning: 0,
+                highest_finalized_num,
+                session_boundaries,
+                client,
+                block_requester,
+                config,
+                messages_from_network,
+                messages_for_aleph,
+            },
+            SimpleNetwork::new(messages_from_data_store, messages_to_network),
+        )
+    }
+
+    pub async fn run(&mut self, mut exit: oneshot::Receiver<()>) {
+        let mut maintenance_clock = Delay::new(self.config.periodic_maintenance_interval);
+        let mut import_stream = self.client.import_notification_stream();
+        let mut finality_stream = self.client.finality_notification_stream();
+        loop {
+            self.prune_pending_messages();
+            self.prune_triggers();
+            tokio::select! {
+                Some(message) = async {
+                    let mut lock = self.messages_from_network.lock().await;
+                    lock.next().await
+                } => {
+                    trace!(target: "afa", "Received message at Data Store {:?}", message);
+                    self.on_message_received(message);
+                }
+                Some(block) = &mut import_stream.next() => {
+                    trace!(target: "afa", "Block import notification at Data Store for block {:?}", block);
+                    self.on_block_imported(block.header);
+                },
+                Some(block) = &mut finality_stream.next() => {
+                    trace!(target: "afa", "Finalized block import notification at Data Store for block {:?}", block);
+                    self.on_block_finalized((block.header.hash(), *block.header.number()).into());
+                }
+                _ = &mut maintenance_clock => {
+                    self.run_maintenance();
+                    maintenance_clock = Delay::new(self.config.periodic_maintenance_interval);
+                }
+                _ = &mut exit => {
+                    debug!(target: "afa", "Data Store task received exit signal. Terminating.");
+                    break;
+                }
+            }
+        }
+    }
+    // Updates our highest known and highest finalized block info directly from the client.
+    fn update_highest_blocks(&mut self) {
+        let client_info = self.client.info();
```

Comment:
Shoudn't this rather be using the `ChainInfoProvider`, since it also has this capability and does (almost?) the same?

Reply:
Yes, certainly. I missed it :/

Reply:
You can remove the `if` around the second call, since it's in `on_finalized` anyway.